ch_pipeline/processing/daily.py

```diff
@@ -406,21 +408,103 @@ def _load_hook(self):
+
+        Total file count is verified by checking files that exist everywhere.
+
+        Parameters
+        ----------
+        start_csd : int
+            Start date in sidereal day format
+        end_csd : int
+            End date in sidereal day format
+
+        Returns
+        -------
+        list
+            List contains the chimestack files available in the timespan, if all of them are available online
+
+        """
+
+        # Connect to databases
+        db.connect()
+
+        # Get timestamps in unix format
+        # Needed for queries
+        start_time = ephemeris.csd_to_unix(start_csd)
+        end_time = ephemeris.csd_to_unix(end_csd)
+
+        # We will want to know which files are in chime_online and nearline on cedar
+        online_node = di.StorageNode.get(name="cedar_online", active=True)
+        chimestack_inst = di.ArchiveInst.get(name="chimestack")
+
+        # TODO if the time range is so small that it’s completely contained within a single file, nothing will be returned
+        # have to special-case it by looking for files which start before the start time and end after the end time).
+
+        archive_files = (
+            di.ArchiveFileCopy.select(
+                di.ArchiveFileCopy.node,
+                di.ArchiveAcq.name.alias("acq_name"),
+                di.ArchiveFile.name.alias("file_name"),
+                di.CorrFileInfo.start_time,
+                di.CorrFileInfo.finish_time,
+            )
+            .join(di.ArchiveFile)
+            .join(di.ArchiveAcq)
+            .switch(di.ArchiveFile)
+            .join(di.CorrFileInfo)
+        )
+
+        # chimestack files available online which include between start and end_time
+
+        files_that_exist = archive_files.where(
+            di.ArchiveAcq.inst
+            == chimestack_inst,  # specifically looking for chimestack files
+            di.CorrFileInfo.start_time
+            < end_time,  # which contain data that includes start time and end time
+            di.CorrFileInfo.finish_time >= start_time,
+            di.ArchiveFileCopy.has_file == "Y",
+        )
+
+        files_online = files_that_exist.where(
+            di.ArchiveFileCopy.node == online_node,  # that are online
+        )
+
+        ChimestackFile = namedtuple("ChimestackFile", "name start_time finish_time")
+
+        filenames_online = [
+            ChimestackFile(os.path.join(acq, name), start_time, finish_time)
+            for node, acq, name, start_time, finish_time in files_online.tuples()
+        ]
+
+        # files_that_exist might contain the same file multiple files
+        # if it exists in multiple locations (nearline, online, gossec, etc)
+        # we only want to include it once
+        filenames_that_exist = []
+        for node, acq, name, start_time, finish_time in files_that_exist.tuples():
+            if not filenames_that_exist or filenames_that_exist[
+                -1
+            ].name != os.path.join(acq, name):
```

Comment:
I think what you are trying to do here is remove the duplicated entries, but I fear it's not very robust as it assumes that the same file names will appear consecutively and I don't see how that is actually guaranteed.

I think there's two options here:

- Re-write the query such that it is based on ArchiveFile's not ArchiveFileCopy's. That would guarantee uniquely returned files. However, that might be a bit of work.
- Use a Python container that will guarantee uniqueness (e.g. a set), and dump anything other than the start and end times from the query. This would should be easy.
```python
ChimestackFile = namedtuple("ChimestackFile", "start_time finish_time")
filenames_that_exist = set(
    ChimestackFile(start_time, finish_time)
    for start_time, finish_time in files_online.tuples()
)
filesnames_that_exist = sorted(list(filenames_that_exist))
```

Reply:
I made this change because `set()` was not working for me for some reason. I now see that it was not working because `set()`s are un-ordered, and it shifted the order. The `sorted()` should fix that. I will make the change, this is better.

Reply:
Bonus: this is so much faster!